keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,423 @@
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [API Changes](#api-changes)
+            * [Container Restart Policy](#container-restart-policy)
+            * [Pod Resize Retry Policy](#pod-resize-retry-policy)
+            * [CRI Changes](#cri-changes)
+         * [Flow Control](#flow-control)
+            * [Transitions of the PodConditions](#transitions-of-the-podconditions)
+            * [Container resource limit update ordering](#container-resource-limit-update-ordering)
+            * [Container resource limit update failure handling](#container-resource-limit-update-failure-handling)
+            * [Notes](#notes)
+         * [Affected Components](#affected-components)
+         * [Possible Extensions](#possible-extensions)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+      * [Alternatives](#alternatives)
+
+## Summary
+
+This proposal aims at allowing Pod resource requests & limits to be updated
+in-place, without a need to restart the Pod or its Containers.
+
+The **core idea** behind the proposal is to make PodSpec mutable with regards to
+Resources, denoting **desired** resources.
+Additionally, PodStatus is extended to provide information about **actual**
+resource allocation.
+
+This document builds upon [proposal for live and in-place vertical scaling][] and
+[Vertical Resources Scaling in Kubernetes][].
+
+[proposal for live and in-place vertical scaling]: https://github.com/kubernetes/community/pull/1719
+[Vertical Resources Scaling in Kubernetes]: https://docs.google.com/document/d/18K-bl1EVsmJ04xeRq9o_vfY2GDgek6B6wmLjXw-kos4/edit?ts=5b96bf40
+
+## Motivation
+
+Resources allocated to a Pod's Container can require a change for various reasons:
+* load handled by the Pod has increased significantly and current resources are
+  not sufficient,
+* load has decreased significantly and allocated resources are unused and wasted,
+* resources have simply been set improperly.
+
+Currently, changing resource allocation requires the Pod to be recreated since
+the PodSpec's Container Resources is immutable.
+
+While many stateless workloads are designed to withstand such a disruption, some
+are more sensitive, especially when using low number of Pod replicas.
+
+Moreover, for stateful or batch workloads, a Pod restart is a serious
+disruption, resulting in lower availability or higher cost of running.
+
+Allowing Resources to be changed without recreating the Pod or restarting the
+Containers addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource resize is not possible.
+* Secondary: allow users to specify which Pods and Containers can be resized
+  without a restart.
```

Comment:
Why do pods need to be able to choose this ?

Reply:
thinking out loud, decreasing the memory limit makes more sense to require a restart. why would increasing the memory limit require a restart ?

Reply:
As mentioned in the use-case, if the java app was started with mxN flag where N is the container Resources.Limit, and it needs more memory at some point, such an app would need a restart even if N were increased in-place. This is not a pod restart, rather a restart of that specific container.

Reply:
A little off topic, but Java now has support for using container resources as limits instead of Xms, Xmx. Not sure if a JVM restart would still be necessary if cgroup limits change. https://github.com/GoogleContainerTools/distroless/issues/293

Reply:
> A little off topic, but Java now has support for using container resources as limits instead of Xms, Xmx. Not sure if a JVM restart would still be necessary if cgroup limits change. [GoogleContainerTools/distroless#293](https://github.com/GoogleContainerTools/distroless/issues/293)

Does this help legacy applications that may still be using Xmx?